frontend/gateway/pb/caps.go

```diff
@@ -15,6 +15,7 @@ const (
 	CapReadFile          apicaps.CapID = "readfile"
 	CapReturnResult      apicaps.CapID = "return"
 	CapReturnMap         apicaps.CapID = "returnmap"
+	//TBD CapClientGateway?
```

Comment:
I think this is probably needed, an older `buildkitd` would successfully return an empty `Result` if `Solve` was called by a new client with `def` and `frontend` both empty, I think the client needs return an error if this is attempted.

Reply:
> make it so that gateway calls to solve continue to return an empty result

Under which circumstances?

Also not sure if by "gateway calls to solve" you mean calls to `controlapi.Solve` with `def == nil && frontend == nil` or when the callback given to `controlapi.Build` calls the `gateway/client.Client.Solve`.

Reply:
The latter. When frontend (or this new logic) calls `Solve()` through the gateway grpc api.

Reply:
I'll add a test case.

Reply:
It `panics` on a nil deref at the `range creq.Definition.Metadata` at the top of `grpcClient.Solve`. Which I suspect a `frontend` could hit today. I'll add the surrounding check along with the test of that code path via this new logic.

Is there scaffolding for integration testing the APIs available to a proper frontend somewhere I'm failing to find? I think that might be a bit hard to arrange (perhaps it would involve `gateway-devel` somehow) but testing via this new mode should suffice for lots of the interesting cases I guess.

Reply:
Yes, the integration tests are lacking that atm. First thing we could do is add a gateway switch to dockerfile tests similarly to the current worker ones. So these tests run using builtin frontend, using the dockerfile pkg directly through client gw and using the master build gateway image. After that we can add something that builds the gateway image from source and tests it directly.

For direct testing of the gateway API though client we can add regular new integration tests under client like you have already started.

Reply:
> So these tests run using builtin frontend, using the dockerfile pkg directly through client gw and using the master build gateway image

So we'd be looking at adding another dimension similar to the code in `util/testutil/integration/run.go` (which runs a set of tests on a range of workers) to `frontend/dockerfile/dockerfile_test.go`, except this time iterating over "modes". I'll have a dig.

I'm going to nuke the
```
 	//TBD CapClientGateway? 
```
comment this thread is derived from (the topic has skewed a bit), which sadly means this will all get collapsed...